django/db/models/sql/query.py

```diff
@@ -1217,17 +1217,15 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
             raise FieldError("Cannot parse keyword query %r" % arg)
         lookups, parts, reffed_expression = self.solve_lookup_type(arg)
 
-        if not getattr(reffed_expression, 'filterable', True):
-            raise NotSupportedError(
-                reffed_expression.__class__.__name__ + ' is disallowed in '
-                'the filter clause.'
-            )
-
         if not allow_joins and len(parts) > 1:
             raise FieldError("Joined field references are not permitted in this query")
 
         pre_joins = self.alias_refcount.copy()
         value = self.resolve_lookup_value(value, can_reuse, allow_joins, simple_col)
+
+        if not getattr(reffed_expression, 'filterable', True) or not getattr(value, 'filterable', True):
+            raise NotSupportedError('Window is disallowed in the filter clause.')
```

Comment:
I think the original `reffed_expression.__class__.__name__` should be used.

Reply:
Imo it can't be used because error message would become really misleading. E.g. `CombinedExpression is disallowed in the filter clause.` - that's simply a false statement. 

I thought of solving this by propagating the original class that causes final expression 
to be non filterable, but decided that it's too heavyweight while only Window fits this description...

Reply:
Hadn't considered this.

Reply:
`filterable` is a documented attribute that can be used by custom expressions, that's why we cannot return  `'Window is disallowed in the filter clause.'` in all cases. IMO we have two choices:

- 1st: use `repr(expression)` instead of `Window` in a message, or
- 2nd: add `for_filter=False` attribute to the `resolve_expression()`  and raise error for non filterable sub-expressions.

Reply:
or 3rd option: we can add `check_filterable()` to `Expression` that will raise exception when expression in not filterable and use it in `build_filter()` on `reffed_expression` and `value`.

Reply:
@felixxm, 1st option means a change in exception message, do you think it's OK? Also do you think `filterable` should be deprecated if/when #28333 is implemented?